syft/messaging/plan.py

```diff
@@ -363,190 +482,82 @@ def build(self, *args):
         Args:
             args: Input data.
         """
-        args = list(args)
 
         # Move the arguments of the first call to the plan and store their ids
-        # as they will be included in the readable_plan: it should be updated
+        # as they will be included in the commands: it should be updated
         # when the function is called with new args and that's why we keep the
-        # refs self.arg_ids
-        self.arg_ids = list()
-        local_args = list()
+        # refs self.procedure.arg_ids
+        arg_ids = list()
+        build_args = list()
         for arg in args:
             arg = arg.send(self)
-            self.arg_ids.append(arg.id_at_location)
-            local_args.append(arg)
+            arg_ids.append(arg.id_at_location)
+            build_args.append(arg)
+        self.procedure.arg_ids = tuple(arg_ids)
 
         # Same for the state element: we send them to the plan and keep reference
-        # to the ids
-        copied_state = self.state.copy()
-        self.state.send(location=self)
-        self.state_ids = self.state.get_id_at_location()
+        # to the remote ids
+        cloned_state = self.state.clone_state()
+        build_state_ids = tuple(self.state.state_ids)  # tuple makes a copy
+        self.state.send_for_build(location=self)
 
         # We usually have include_state==True for functions converted to plan
         # using @func2plan and we need therefore to add the state manually
         if self.include_state:
-            res_ptr = self.forward(*local_args, self.state)
+            res_ptr = self.forward(*build_args, self.state)
         else:
-            res_ptr = self.forward(*local_args)
-        res_ptr.child.garbage_collect_data = False
+            res_ptr = self.forward(*build_args)
 
-        # We put back the original state. There is no need to keep a reference
-        # To the element sent to the plan.
-        self.state.set_(copied_state)
+        # We put back a copy of the original state
+        self.state.set_(cloned_state)
+        # so we update the procedure
+        self.procedure.update_ids(from_ids=build_state_ids, to_ids=self.state.state_ids)
 
         # The readable plan is now built, we hide the fact that it was run on
```

Comment:
The term readable plan has been replaced

Reply:
indeed